[clang] `-fpch-debuginfo` suppresses vtable emission, causing link errors ("undefined symbol: vtable for X")

## Summary

When a class with a vtable is defined in a header compiled with `-fpch-debuginfo`, clang sometimes fails to emit the vtable in the key-function translation unit. Any other TU that instantiates the class then fails to link:

```
Undefined symbols for architecture arm64:
  "vtable for B", referenced from:
      B::B() in main.o
```

This is a regression introduced by commit 847f9cb0e868.

## Reproducer

`b.h`:
```cpp
#pragma once
struct B {
    B() = default;
    virtual ~B();  // key function
    virtual void f();
};
```

`b.cpp` (key-function TU):
```cpp
#include "b.h"
B::~B() {}
void B::f() {}
```

`main.cpp`:
```cpp
#include "b.h"
int main() {
    B b;
    b.f();
}
```

```sh
# Build PCH with -fpch-debuginfo
clang++ -std=c++20 -x c++-header b.h -o b.h.pch -fpch-debuginfo

# Compile both TUs against the PCH
clang++ -std=c++20 -c b.cpp -o b.o -include-pch b.h.pch
clang++ -std=c++20 -c main.cpp -o main.o -include-pch b.h.pch

# Link -- fails on affected builds
clang++ b.o main.o -o a.out
```

--- a/clang/include/clang/AST/DeclBase.h
+++ b/clang/include/clang/AST/DeclBase.h
@@ -687,6 +687,10 @@ public:
 
   /// Whether the definition of the declaration should be emitted in external
   /// sources.
+  /// FIXME: This conflates two questions: if the entity should be emitted into
+  ///   other object files (because there's no primary), and if the debug info
+  ///   should be emitted into other object files. This matters for
+  //    `-fmodules-debuginfo`, `-fmodules-codgen`, and `isInNamedModule()`.
   bool shouldEmitInExternalSource() const;
 
   /// Whether this declaration comes from explicit global module.

--- a/clang/lib/Sema/SemaDeclCXX.cpp
+++ b/clang/lib/Sema/SemaDeclCXX.cpp
@@ -19296,7 +19296,12 @@ bool Sema::DefineUsedVTables() {
     DefinedAnything = true;
     MarkVirtualMembersReferenced(Loc, Class);
     CXXRecordDecl *Canonical = Class->getCanonicalDecl();
-    if (VTablesUsed[Canonical] && !Class->shouldEmitInExternalSource())
+    // The vtable is assumed to be emitted in an external source only for
+    // classes attached to a named module, which is guaranteed to have an object
+    // file. This isn't true for -fmodules-debuginfo, which still has
+    // shouldEmitInExternalSource as true so that debug info gets supressed.
+    if (VTablesUsed[Canonical] &&
+        !(Class->isInNamedModule() && Class->shouldEmitInExternalSource()))
       Consumer.HandleVTable(Class);
 
     // Warn if we're emitting a weak vtable. The vtable will be weak if there is

--- a/clang/test/PCH/pch-debuginfo-vtable.cpp
+++ b/clang/test/PCH/pch-debuginfo-vtable.cpp
@@ -1,0 +1,28 @@
+// Check that a key function defined in a TU that includes a -fpch-debuginfo
+// PCH still emits the vtable. The PCH object only carries debug info, not the
+// vtable definition, so the importing TU must emit it itself.
+
+// RUN: rm -rf %t
+// RUN: mkdir -p %t
+// RUN: split-file %s %t
+
+// RUN: %clang_cc1 -std=c++20 -triple %itanium_abi_triple -fmodules-debuginfo \
+// RUN:     -building-pch-with-obj -x c++-header -emit-pch %t/b.h -o %t/b.pch
+
+// RUN: %clang_cc1 -std=c++20 -triple %itanium_abi_triple -include-pch %t/b.pch \
+// RUN:     -emit-llvm -o - %t/b.cpp | FileCheck %s
+
+// CHECK: @_ZTV1B = {{.*}}constant
+
+//--- b.h
+#pragma once
+struct B {
+    B() = default;
+    virtual ~B();
+    virtual void f();
+};
+
+//--- b.cpp
+#include "b.h"
+B::~B() { }
+void B::f() { }

